satpy/demo/seviri_hrit.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021 Satpy developers
+#
+# This file is part of satpy.
+#
+# satpy is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Demo data download for SEVIRI HRIT files."""
+
+import logging
+import os.path
+
+from satpy import config
+from satpy.demo.utils import download_url
+
+logger = logging.getLogger(__name__)
+
+ZENODO_BASE_URL = "https://zenodo.org/api/files/dcc5ab29-d8a3-4fb5-ab2b-adc405d18c23/"
+FILENAME = "H-000-MSG4__-MSG4________-{channel:_<9s}-{segment:_<9s}-201802281500-__"
+
+
+def download_seviri_hrit_20180228_1500(base_dir=None, subset=None):
+    """Download the SEVIRI HRIT files for 2018-02-28T15:00.
+
+    *subset* is a dictionary with the channels as keys and granules to download
+    as values, eg::
+
+      {"HRV": [1, 2, 3], "IR_108": [1, 2], "EPI": None}
+
+    """
+    files = generate_subset_of_filenames(subset)
+
+    base_dir = base_dir or config.get("demo_data_dir", ".")
+    subdir = os.path.join(base_dir, "seviri_hrit", "20180228_1500")
+    os.makedirs(subdir, exist_ok=True)
+    targets = []
+    for the_file in files:
+        target = os.path.join(subdir, the_file)
+        targets.append(target)
+        if os.path.isfile(target):
+            continue
+        download_url(ZENODO_BASE_URL + the_file, target)
+    return targets
+
+
+def generate_subset_of_filenames(subset=None, base_dir=""):
+    """Generate SEVIRI HRIT filenames."""
+    if subset is None:
+        subset = _create_full_set()
+    pattern = os.path.join(base_dir, FILENAME)
+    files = []
+    for key, value in subset.items():
+        if key in ["PRO", "EPI"]:
+            files.append(pattern.format(channel="", segment=key))
+        else:
+            for segment in value:
+                files.append(pattern.format(channel=key, segment=f"{segment:06d}"))
```

Comment:
You're more "clean code"-y than I am, but this seems like a lot of levels of indentation.

Reply:
Fixed